feat windivert install workaround, assets, separate module for constants

averageeucplayer opened the pull request.

snoww commented:
will the rest of your prs work if #162 gets merged?

anyduck commented on src-tauri/src/misc.rs:
```diff
@@ -0,0 +1,33 @@
+///
+/// ### Windows-specific
+/// This function is only relevant on Windows platforms. It is a workaround for cases where
+/// the DLL (`WinDivert.dll`) and SYS (`WinDivert64.sys`) driver files are locked by an already running kernel driver.
```
This comment seems wrong. When WinDivert is locked, these files still exist on the fs. Looks to me more like a case of detecting when DLLs are quarantined

anyduck commented on src-tauri/src/misc.rs:
```diff
@@ -0,0 +1,33 @@
+        if !windivert_driver_path.exists() {
+            use std::fs;
+
+            let bytes: &'static [u8] = include_bytes!("../WinDivert64.sys");
```
Will AVs now quarantine whole LOA logs binary instead of dlls? I wanna test it, but idk which one are triggered by windivert

snoww replied:
good point. I wonder if that was the case, it would make the user's life easier by letting them whitelist the app, instead of them needing to whitelist the driver, or will it cause more headache as people will actually think the meter itself is malicious and become hesitant to install

averageeucplayer commented:
> will the rest of your prs work if #162 gets merged?

I predict there will be some conflicts on the ui side, and in main.rs

averageeucplayer commented:
done

snoww commented:
instead of using the crate `once_cell`, can't we utilize the stablized `OnceCell` in std? https://doc.rust-lang.org/std/cell/struct.OnceCell.html

snoww commented:
also I'm not the biggest fan of using unsafe in `data.rs`

averageeucplayer commented:
- Added the missing JSON file
- Switched once_cell:sync::Lazy to [std::sync::LazyLock](https://doc.rust-lang.org/std/sync/struct.LazyLock.html)
- Upgraded rustc/tauri to 1.89/2.8.4

Ran into some odd errors because tauri.conf.json was being parsed using a cached old schema. Fixed it with:

```sh
cargo tauri uninstall
cargo install tauri-cli --version "^2.8.3"--locked
```

Smoke tested app ( ie click throw the tabs, modify settings ), no issues so far

averageeucplayer commented:
Yeah, forgot about BufReader. reading 8kb chunks while it has process over 80mb is definitely not efficient

averageeucplayer commented:
Even if we end up needing the whole dataset, baking it into the binary with include_str!/include_bytes! still has some trade-offs:

The bytes live forever in the .rdata section, and then we allocate a second copy in heap after parsing, so memory use is higher than just loading and discarding.

The OS loader has to map the entire blob at program start, which slows startup compared to deferring the read until we need it.

Build times get heavier since the compiler and linker carry around the extra 80 MB every build.

Data updates require a full rebuild, whereas with an external file we could ship a new dataset without touching the binary. ( perhaps a nice feature to implement in the future )

snoww commented:
thats a good point. I also read that deserializing from a buf can be slower than just reading the string. what are your thoughts on that? https://github.com/serde-rs/json/issues/160

anyduck commented on src-tauri/src/context.rs:
```diff
@@ -0,0 +1,41 @@
+
+impl AppContext {
+    pub fn new() -> Result<Self> {
+        let instance = SingleInstance::new("LOA_LOGS").unwrap();
+        ensure!(instance.is_single(), "Another instance of {} is already running!", "LOA Logs");
```
There's already plugin-single-instance. If you need to customize behavior, you can do that in the function passed to `tauri_plugin_single_instance::init`
https://github.com/snoww/loa-logs/blob/83e951f6c7208e0b71097119efcad14cba01ad11/src-tauri/src/main.rs#L56

averageeucplayer replied:
I’m not a fan of this plugin, you have to wait for the entire Tauri infrastructure to spin up just to be told the app is already running.

anyduck commented on src-tauri/src/main.rs:
```diff
@@ -27,19 +30,16 @@ use tauri_plugin_updater::UpdaterExt;
-);
+use crate::constants::*;
+use crate::context::AppContext;
+use crate::live::data::AssetPreloader;
```
Can't compile without the `meter-core` flag because `crate::live` is used here

averageeucplayer commented on src-tauri/src/live/data.rs:
```diff
@@ -0,0 +1,112 @@
+    let file = File::open(path).unwrap_unchecked();
+    let reader = BufReader::with_capacity(1024 * 1024, file);
+    serde_json::from_reader(reader).unwrap_unchecked()
+}
```
done

anyduck commented:
Ah, I need to request changes instead of commenting for that to work

averageeucplayer commented:
just noticed small config sneaked in if you don't mind

```
[profile.dev]
incremental = true
opt-level = 0
```

should be by default in recent cargo edition

molenzwiebel commented:
There's also [uneval](https://crates.io/crates/uneval) as a secret bonus option, but I expect that that might result in lovely compile times when you apply it to the skill effects :)

snoww commented:
still have the issue of windivert drivers being left behind when uninstalling the app, due to it not being included in the install by default

averageeucplayer commented:
testing uninstall with following tweak in nsi

```
Delete /REBOOTOK "$INSTDIR\WinDivert64.sys"
```

seems to work.
